docs/pallas/tpu/pipelining.md

```diff
@@ -0,0 +1,397 @@
+
++++ {"id": "5kWr-1tKpYro"}
+
+## Constraints of using VMEM/SMEM
+
+Pallas exposes access to lower level memory spaces like VMEM and SMEM but writing kernels utilizing them adds some considerations.
+
+1. Memory capacity. VMEM and SMEM are *small*! VMEM on v4 TPUs is only 16MiB and SMEM ranges in the tens to hundreds of KiB. If our arrays are too big, we won't even be able to fit them into VMEM at all. For reference, a `f32[2048, 2048]` array is 16MiB, so our above kernel won't scale beyond moderately sized arrays.
+
+2. Memory bandwidth. Copying to/from HBM and VMEM takes a long time, at least compared to most compute instructions. The `add_matrices` function above will likely spend more time copying between HBM and VMEM than actually performing the addition itself.
+
+With these two constraints in mind, we'll have to rethink our strategy for getting performance of our TPUs.
+
++++ {"id": "_NTqvlbetB3P"}
+
+## Primer: Pipelining
+
+Pipelining our computation offers a way of dealing with both the memory capacity and bandwidth constraints in one fell swoop. What do we mean by pipelining?
+
+The goal is: *concurrently* copy to/from HBM and VMEM *while* utilizing our compute units. However, the way we expressed our program above is that we copy *all* of `x` and `y` before we start doing any compute with them. This creates a dependence between the copy and the compute.
+
+However, if we can chunk up our computation into several subcomputations (e.g. when we add two matrices, we can express that as addition of "blocks" of the original matrices together), we can now overlap the copies of one of those subcomputations with the compute of the other. Let's walk through a simple example:
+
+Let's say we split our arrays `x` and `y` into `x1, x2` and `y1, y2` (for example, split along the leading axis, resulting in two `(256, 512)` arrays for each input. We can now execute the following pipelined computation.
+
+1. Copy `x1` and `y1` into VMEM.
+1. Start copying `x2` and `y2` into VMEM
+2. Load `x1, y1` from VMEM into VREGs.
+3. Execute the `z1 = x1 + y1` using the compute units.
+4. Store `z1` into VMEM.
+5. Start copying `z1` from VMEM back into HBM.
+6. Wait until `x2, y2` have been copied into VMEM.
+7. Load `x2, y2` from VMEM into VREGs.
+8. Execute the `z2 = x2 + y2` using the compute units.
+9. Store `z2` into VMEM.
+10. Wait until `z1` is copied into HBM.
+10. Start copying `z2` from VMEM back into HBM.
+10. Wait until `z2` is copied into HBM.
+
+Any time we are doing compute here, we are asynchronously copying something. This means that some of the time spent copying is not wasted.
+
+The two most important numbers for determining how efficient a pipelined computation are a) how many FLOP/s we can execute and b) how many bytes/s we can copy. The ratio of these two (FLOPs/bandwidth) is called the *arithmetic intensity* of an operation.
```

Comment:
I don't think this is consistent with [the wikipedia definition of arithmetic intensity](https://en.wikipedia.org/wiki/Roofline_model#:~:text=Arithmetic%20intensity,-The%20arithmetic%20intensity&text=will%20be%20the%20ratio%20of,movement%20(FLOPs%2Fbyte).).  That definition is about FLOPs (ie total number of floating point operations performed, not FLOPS or FLOP/s, ie floating point operations per second) compared to total bytes of traffic (ie not bytes/sec). Also I think those quantities are properties of a workload, e.g. matrix multiplication of particular shapes, rather than a chip's FLOP/s and bytes/s. See also [this ref](https://people.eecs.berkeley.edu/~kubitron/cs252/handouts/papers/RooflineVyNoYellow.pdf).

Maybe there are some other definitions of the term 'arithmetic intensity' that are consistent with what's written here, but if so can you link them? If not, let's adapt to be consistent with Wikipedia and link that. Or just delete this paragraph because I don't think arithmetic intensity is ever referenced again in the doc.

Reply:
Fixed.